sagemathinc/cocalc
# Jupyter explicit kernel

# Description
Jupyter notebooks without a kernel or an unknown one are a source of confusion. This patch presents a clear selection dialog with some explanations about what a kernel actually is to the user.

**Reference**: there are two situations the dialog shows up. either user requested, or when opening and the kernel is "bad". here is a side-by-side view of running this in two instances. The dialog only closes  when a kernel is set by itself, when it wasn't opened by the user in the first place.

* [no kernel](https://share.cocalc.com/share/14eed217-2d3c-4975-a381-b69edcb40e0e/scratch/no-kernel.webm?viewer=share)
* [unknown kernel](https://share.cocalc.com/share/14eed217-2d3c-4975-a381-b69edcb40e0e/scratch/unknown-kernel.webm?viewer=share)

# Testing Steps
**important:** switch to experimental image to see the suggestions. 

1. existing notebook with known kernel → this doesn't show up
1. click on kernel info at the top right → kernel "change" dialog shows up
1. open a new ipynb file → dialog asks what to do (slightly different from above)
1. an unknown kernel can be created like that:

```
~$ cat <<EOF > unknown-7.ipynb
{
 "cells": [ ],
 "metadata": {
  "kernelspec": {
   "name": "pythonXY"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
EOF

~$ open unknown-7.ipynb
```

# Relevant Issues
I didn't check

### [Checklist](https://github.com/sagemathinc/cocalc/wiki/PR-Checklist):

Front end:

Back end:
- [x] it touches the "project"!
